Remove CDN signature before hashing asset URL

slatinsky opened the pull request:
Discord recently introduced signed cdn links. This change breaks `--reuse-media` for assets, because url is used to detect if the file is already downloaded.

This pull request solves this issue by stripping out `ex`, `is` and `hm` query params before hashing the url.

Tyrrrz commented on DiscordChatExporter.Core/Exporting/ExportAssetDownloader.cs:
```diff
@@ -101,12 +102,24 @@ await using (var output = File.Create(filePath))
 {
-    private static string GetUrlHash(string url) =>
-        SHA256
+    private static string GetUrlHash(string url)
+    {
+        // Strip out ex, is and hm query params to create a consistent hash
+        var uriBuilder = new UriBuilder(url);
+        var query = HttpUtility.ParseQueryString(uriBuilder.Query);
+        query.Remove("ex");
+        query.Remove("is");
+        query.Remove("hm");
+        uriBuilder.Query = query.ToString();
+        url = uriBuilder.ToString();
+        url = url.Replace(":443", "");
+
+        return SHA256
```
Let's do it only for Discord CDN URLs. These parameters can have some other meaning on other hosts.

slatinsky commented:
check for discord cdn added. Also added (simple) check if the url is really signed by checking for `ex` `is` and `hm` substrings (could be improved with regex if needed). For example avatars and emojis are not signed.

Also fixed formatting error introduced by the previous PR.

Tyrrrz commented on DiscordChatExporter.Core/Exporting/ExportAssetDownloader.cs:
```diff
@@ -101,12 +102,32 @@ await using (var output = File.Create(filePath))
-        SHA256
+    private static string GetUrlHash(string url)
+    {
+        // Strip out ex, is and hm query params from Discord CDN URLs to create a consistent hash
+        if (
+            url.Contains("cdn.discordapp.com")
+            && url.Contains("ex")
+            && url.Contains("is")
+            && url.Contains("hm")
+        )
+        {
+            var uriBuilder = new UriBuilder(url);
+            var query = HttpUtility.ParseQueryString(uriBuilder.Query);
+            query.Remove("ex");
+            query.Remove("is");
+            query.Remove("hm");
+            uriBuilder.Query = query.ToString();
+            url = uriBuilder.ToString();
+            url = url.Replace(":443", "");
+        }
```
```suggestion
        // Strip out signature parameters from Discord CDN URLs
        if (url.Contains("cdn.discordapp.com"))
        {
            var query = HttpUtility.ParseQueryString(uriBuilder.Query);
            query.Remove("ex");
            query.Remove("is");
            query.Remove("hm");

            url = new UriBuilder(url)
            {
                Query = query.ToString()
            }.ToString();
        }
```

Tyrrrz commented on DiscordChatExporter.Core/Exporting/ExportAssetDownloader.cs:
```diff
@@ -101,12 +102,32 @@ await using (var output = File.Create(filePath))
+            query.Remove("hm");
+            uriBuilder.Query = query.ToString();
+            url = uriBuilder.ToString();
+            url = url.Replace(":443", "");
```
Is this required? 🤔

slatinsky replied:
```cs
url = url.Replace(":443", "");
```
this is needed, because `Uribuilder` for some reason adds a port to the url
`https://cdn.discordapp.com/attachments/...` -> `https://cdn.discordapp.com:443/attachments/...`

The goal is to transform the url to the format as the one before Discord added signed urls.

Tyrrrz commented:
Thanks!
